diff -Naur libarchive-54/config.h libarchive/config.h
--- libarchive-54/config.h	2012-04-17 04:07:27.000000000 +0200
+++ libarchive/config.h	2017-12-05 18:03:29.000000000 +0100
@@ -222,7 +222,7 @@
 /* #undef HAVE_LIBEXPAT */
 
 /* Define to 1 if you have the `lzma' library (-llzma). */
-#define HAVE_LIBLZMA 1
+// #define HAVE_LIBLZMA 1
 
 /* Define to 1 if you have the `lzmadec' library (-llzmadec). */
 /* #undef HAVE_LIBLZMADEC */
@@ -274,7 +274,7 @@
 /* #undef HAVE_LZMADEC_H */
 
 /* Define to 1 if you have the <lzma.h> header file. */
-#define HAVE_LZMA_H 1
+// #define HAVE_LZMA_H 1
 
 /* Define to 1 if you have the `MD5Init' function. */
 /* #undef HAVE_MD5INIT */
diff -Naur libarchive-54/libarchive/libarchive/archive_read_support_format_iso9660.c libarchive/libarchive/libarchive/archive_read_support_format_iso9660.c
--- libarchive-54/libarchive/libarchive/archive_read_support_format_iso9660.c	2012-09-21 00:45:16.000000000 +0200
+++ libarchive/libarchive/libarchive/archive_read_support_format_iso9660.c	2017-12-05 18:06:13.000000000 +0100
@@ -2678,7 +2678,7 @@
 		parent_key = heap->files[parent]->key;
 		if (file_key >= parent_key) {
 			heap->files[hole] = file;
-			return;
+			return 0; // Non-void function should return a value
 		}
 		// Move parent into hole <==> move hole up tree.
 		heap->files[hole] = heap->files[parent];
diff -Naur libarchive-54/libarchive/libarchive/archive_util.c libarchive/libarchive/libarchive/archive_util.c
--- libarchive-54/libarchive/libarchive/archive_util.c	2010-05-21 03:07:32.000000000 +0200
+++ libarchive/libarchive/libarchive/archive_util.c	2017-12-05 20:42:57.000000000 +0100
@@ -40,6 +40,7 @@
 #include "archive.h"
 #include "archive_private.h"
 #include "archive_string.h"
+#include "ios_error.h"  // remaps exit() to pthread_exit(NULL)
 
 #if ARCHIVE_VERSION_NUMBER < 3000000
 /* These disappear in libarchive 3.0 */
diff -Naur libarchive-54/libarchive/libarchive/archive_write_set_format_pax.c libarchive/libarchive/libarchive/archive_write_set_format_pax.c
--- libarchive-54/libarchive/libarchive/archive_write_set_format_pax.c	2010-05-21 03:07:32.000000000 +0200
+++ libarchive/libarchive/libarchive/archive_write_set_format_pax.c	2017-12-05 20:41:58.000000000 +0100
@@ -40,6 +40,7 @@
 #include "archive_entry.h"
 #include "archive_private.h"
 #include "archive_write_private.h"
+#include "ios_error.h"  // remaps exit() to pthread_exit(NULL)
 
 struct pax {
 	uint64_t	entry_bytes_remaining;
diff -Naur libarchive-54/libarchive/libarchive/filter_fork.c libarchive/libarchive/libarchive/filter_fork.c
--- libarchive-54/libarchive/libarchive/filter_fork.c	2010-05-21 03:07:32.000000000 +0200
+++ libarchive/libarchive/libarchive/filter_fork.c	2017-12-05 20:42:13.000000000 +0100
@@ -24,6 +24,7 @@
  */
 
 #include "archive_platform.h"
+#include "ios_error.h"  // remaps exit() to pthread_exit(NULL)
 
 /* This capability is only available on POSIX systems. */
 #if defined(HAVE_PIPE) && defined(HAVE_FCNTL) && \
diff -Naur libarchive-54/libarchive/libarchive_fe/err.c libarchive/libarchive/libarchive_fe/err.c
--- libarchive-54/libarchive/libarchive_fe/err.c	2010-05-21 03:07:32.000000000 +0200
+++ libarchive/libarchive/libarchive_fe/err.c	2017-12-05 20:46:12.000000000 +0100
@@ -38,7 +38,8 @@
 #include <string.h>
 #endif
 
-#include "err.h"
+#include "err.h"   
+#include "ios_error.h"  // remaps exit() to pthread_exit(NULL)
 
 const char *lafe_progname;
 
diff -Naur libarchive-54/libarchive/tar/bsdtar.c libarchive/libarchive/tar/bsdtar.c
--- libarchive-54/libarchive/tar/bsdtar.c	2015-08-20 00:50:01.000000000 +0200
+++ libarchive/libarchive/tar/bsdtar.c	2017-12-05 20:52:03.000000000 +0100
@@ -71,6 +71,7 @@
 
 #include "bsdtar.h"
 #include "err.h"
+#include "ios_error.h"  // remaps exit() to pthread_exit(NULL)
 
 #ifdef __MINGW32__
 int _CRT_glob = 0; /* Disable broken CRT globbing. */
@@ -118,7 +119,7 @@
 	 | ARCHIVE_EXTRACT_SECURE_NODOTDOT)
 
 int
-main(int argc, char **argv)
+tar_main(int argc, char **argv)
 {
 	struct bsdtar		*bsdtar, bsdtar_storage;
 	int			 opt, t;
diff -Naur libarchive-54/libarchive/tar/cmdline.c libarchive/libarchive/tar/cmdline.c
--- libarchive-54/libarchive/tar/cmdline.c	2010-05-21 03:07:32.000000000 +0200
+++ libarchive/libarchive/tar/cmdline.c	2017-12-05 18:18:14.000000000 +0100
@@ -58,7 +58,7 @@
  * a small change to the code below.
  */
 
-static struct option {
+static struct tar_option {
 	const char *name;
 	int required;      /* 1 if this option requires an argument. */
 	int equivalent;    /* Equivalent short option. */
@@ -180,7 +180,7 @@
 	static int state = state_start;
 	static char *opt_word;
 
-	const struct option *popt, *match = NULL, *match2 = NULL;
+	const struct tar_option *popt, *match = NULL, *match2 = NULL;
 	const char *p, *long_prefix = "--";
 	size_t optlength;
 	int opt = '?';
diff -Naur libarchive-54/libarchive/tar/getdate.c libarchive/libarchive/tar/getdate.c
--- libarchive-54/libarchive/tar/getdate.c	2010-05-21 03:07:32.000000000 +0200
+++ libarchive/libarchive/tar/getdate.c	2017-12-05 20:42:47.000000000 +0100
@@ -37,6 +37,7 @@
 #include <stdlib.h>
 #include <string.h>
 #include <time.h>
+#include "ios_error.h"  // remaps exit() to pthread_exit(NULL)
 
 /* This file defines a single public function. */
 time_t get_date(time_t now, char *);
diff -Naur libarchive-54/libarchive/tar/write.c libarchive/libarchive/tar/write.c
--- libarchive-54/libarchive/tar/write.c	2010-05-21 03:07:32.000000000 +0200
+++ libarchive/libarchive/tar/write.c	2017-12-05 20:44:21.000000000 +0100
@@ -89,6 +89,7 @@
 #include <paths.h>
 #endif
 
+#include "ios_error.h"  // remaps exit() to pthread_exit(NULL)
 #include "bsdtar.h"
 #include "err.h"
 #include "line_reader.h"
