diff -Naur shell_cmds-203/date/date.c shell_cmds/date/date.c
--- shell_cmds-203/date/date.c	2015-11-07 23:19:55.000000000 +0100
+++ shell_cmds/date/date.c	2017-12-05 15:41:01.000000000 +0100
@@ -57,13 +57,14 @@
 #include <utmpx.h>
 
 #ifdef __APPLE__
-#include <get_compat.h>
-#else
+// #include <get_compat.h>
+// #else
 #define COMPAT_MODE(a,b) (1)
 #endif /* __APPLE__ */
 
 #include "extern.h"
 #include "vary.h"
+#include "ios_error.h"
 
 #ifndef	TM_YEAR_BASE
 #define	TM_YEAR_BASE	1900
@@ -84,7 +85,7 @@
 static const char *rfc2822_format = "%a, %d %b %Y %T %z";
 
 int
-main(int argc, char *argv[])
+date_main(int argc, char *argv[])
 {
 	struct timezone tz;
 	int ch, rflag;
@@ -324,7 +325,7 @@
 
 	if (!jflag) {
 		/* set the time */
-		if (nflag || netsettime(tval)) {
+		if (nflag /* || netsettime(tval)*/ ) {
 			utx.ut_type = OLD_TIME;
 			(void)gettimeofday(&utx.ut_tv, NULL);
 			pututxline(&utx);
diff -Naur shell_cmds-203/env/env.c shell_cmds/env/env.c
--- shell_cmds-203/env/env.c	2014-08-05 00:47:11.000000000 +0200
+++ shell_cmds/env/env.c	2017-12-05 15:42:21.000000000 +0100
@@ -50,6 +50,7 @@
 #include <unistd.h>
 
 #include "envopts.h"
+#include "ios_error.h"
 
 extern char **environ;
 
@@ -58,7 +59,7 @@
 static void usage(void);
 
 int
-main(int argc, char **argv)
+env_main(int argc, char **argv)
 {
 	char *altpath, **ep, *p, **parg;
 	char *cleanenv[1];
diff -Naur shell_cmds-203/id/id.c shell_cmds/id/id.c
--- shell_cmds-203/id/id.c	2014-02-18 21:44:21.000000000 +0100
+++ shell_cmds/id/id.c	2017-12-05 15:55:18.000000000 +0100
@@ -62,18 +62,19 @@
 #include <stdlib.h>
 #include <string.h>
 #include <unistd.h>
+#include "ios_error.h"
 
-void	id_print(struct passwd *, int, int, int);
-void	pline(struct passwd *);
-void	pretty(struct passwd *);
-void	auditid(void);
-void	fullname(struct passwd *);
-void	group(struct passwd *, int);
-void	maclabel(void);
-void	usage(void);
+static void	id_print(struct passwd *, int, int, int);
+static void	pline(struct passwd *);
+static void	pretty(struct passwd *);
+static void	auditid(void);
+static void	fullname(struct passwd *);
+static void	group(struct passwd *, int);
+static void	maclabel(void);
+static void	usage(void);
 struct passwd *who(char *);
 
-int isgroups, iswhoami;
+static int isgroups, iswhoami;
 
 #ifdef __APPLE__
 // SPI for 5235093
@@ -81,7 +82,7 @@
 #endif
 
 int
-main(int argc, char *argv[])
+id_main(int argc, char *argv[])
 {
 	struct group *gr;
 	struct passwd *pw;
@@ -93,6 +94,9 @@
 	Gflag = Mflag = Pflag = gflag = nflag = pflag = rflag = uflag = 0;
 	Aflag = 0;
 	Fflag = 0;
+    // Initialize:
+    isgroups = 0;
+    iswhoami = 0;
 
 	myname = strrchr(argv[0], '/');
 	myname = (myname != NULL) ? myname + 1 : argv[0];
@@ -483,6 +487,7 @@
 	if (*u && !*ep && (pw = getpwuid(id)))
 		return(pw);
 	errx(1, "%s: no such user", u);
+    return NULL; // keep compiler happy
 	/* NOTREACHED */
 }
 
diff -Naur shell_cmds-203/printenv/printenv.c shell_cmds/printenv/printenv.c
--- shell_cmds-203/printenv/printenv.c	2014-08-05 00:47:11.000000000 +0200
+++ shell_cmds/printenv/printenv.c	2017-12-05 15:44:07.000000000 +0100
@@ -48,8 +48,9 @@
 #include <stdio.h>
 #include <string.h>
 #include <unistd.h>
+#include "ios_error.h"
 
-void	usage(void);
+static void	usage(void);
 extern char **environ;
 
 /*
@@ -59,7 +60,7 @@
  * February, 1979
  */
 int
-main(int argc, char *argv[])
+printenv_main(int argc, char *argv[])
 {
 	char *cp, **ep;
 	size_t len;
@@ -91,7 +92,7 @@
 	exit(1);
 }
 
-void
+static void
 usage(void)
 {
 	(void)fprintf(stderr, "usage: printenv [name]\n");
diff -Naur shell_cmds-203/pwd/pwd.c shell_cmds/pwd/pwd.c
--- shell_cmds-203/pwd/pwd.c	2006-09-12 00:14:13.000000000 +0200
+++ shell_cmds/pwd/pwd.c	2017-12-05 15:44:45.000000000 +0100
@@ -50,12 +50,13 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <unistd.h>
+#include "ios_error.h"
 
 static char *getcwd_logical(void);
-void usage(void);
+static void usage(void);
 
 int
-main(int argc, char *argv[])
+pwd_main(int argc, char *argv[])
 {
 	int physical;
 	int ch;
@@ -97,7 +98,7 @@
 	exit(0);
 }
 
-void
+static void
 usage(void)
 {
 
diff -Naur shell_cmds-203/uname/uname.c shell_cmds/uname/uname.c
--- shell_cmds-203/uname/uname.c	2013-09-26 01:55:54.000000000 +0200
+++ shell_cmds/uname/uname.c	2017-12-05 15:46:31.000000000 +0100
@@ -49,14 +49,15 @@
 
 #include <sys/sysctl.h>
 #include <sys/utsname.h>
+#include "ios_error.h"
 
 #ifdef __APPLE__
-#include <get_compat.h>
-#else  /* !__APPLE__ */
+// #include <get_compat.h>
+// #else  /* !__APPLE__ */
 #define COMPAT_MODE(a,b) (1)
 #endif /* __APPLE__ */
 
-int	main __P((int, char **));
+int	uname_main __P((int, char **));
 static void usage __P((void));
 
 /* Note that PRINT_MACHINE_ARCH is excluded from PRINT_ALL! */
@@ -70,7 +71,7 @@
     (PRINT_SYSNAME|PRINT_NODENAME|PRINT_RELEASE|PRINT_VERSION|PRINT_MACHINE)
 
 int
-main(argc, argv) 
+uname_main(argc, argv) 
 	int argc;
 	char **argv;
 {
diff -Naur shell_cmds-203/w/proc_compare.c shell_cmds/w/proc_compare.c
--- shell_cmds-203/w/proc_compare.c	2007-01-19 02:15:36.000000000 +0100
+++ shell_cmds/w/proc_compare.c	2017-12-05 15:49:38.000000000 +0100
@@ -38,13 +38,15 @@
 __FBSDID("$FreeBSD: src/usr.bin/w/proc_compare.c,v 1.9 2004/04/14 09:34:17 bde Exp $");
 #endif
 
+#define HAVE_UTMPX 1
 #include <sys/param.h>
 #ifdef __APPLE__
 #include <sys/time.h>
 #endif
 #include <sys/proc.h>
 #include <sys/time.h>
-#include <sys/user.h>
+// #include <sys/user.h>
+#include <sys/sysctl.h>
 
 #include "extern.h"
 
diff -Naur shell_cmds-203/w/w.c shell_cmds/w/w.c
--- shell_cmds-203/w/w.c	2007-10-03 11:55:12.000000000 +0200
+++ shell_cmds/w/w.c	2017-12-05 15:52:22.000000000 +0100
@@ -50,15 +50,16 @@
  * This program is similar to the systat command on Tenex/Tops 10/20
  *
  */
+#define HAVE_UTMPX 1
 #include <sys/param.h>
 #include <sys/time.h>
 #include <sys/stat.h>
 #include <sys/sysctl.h>
 #include <sys/proc.h>
-#include <sys/user.h>
+// #include <sys/user.h>
 #include <sys/ioctl.h>
 #include <sys/socket.h>
-#include <sys/tty.h>
+// #include <sys/tty.h>
 
 #ifndef __APPLE__
 #include <machine/cpu.h>
@@ -79,13 +80,13 @@
 #include <limits.h>
 #include <locale.h>
 #include <netdb.h>
-#include <nlist.h>
+#include <mach-o/nlist.h>
 #include <paths.h>
 #include <resolv.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
-#include <timeconv.h>
+// #include <timeconv.h>
 #include <unistd.h>
 #if HAVE_UTMPX
 #include <utmpx.h>
@@ -100,6 +101,7 @@
 #include <TargetConditionals.h>
 
 #include "extern.h"
+#include "ios_error.h"
 
 struct timeval	boottime;
 #if !HAVE_UTMPX
@@ -127,8 +129,8 @@
 /*
  * One of these per active utmp entry.
  */
-struct	entry {
-	struct	entry *next;
+struct	utmp_entry {
+	struct	utmp_entry *next;
 #if HAVE_UTMPX
 	struct	utmpx utmp;
 #else
@@ -161,7 +163,7 @@
 char *fmt_argv(char **, char *, int);	/* ../../bin/ps/fmt.c */
 
 int
-main(int argc, char *argv[])
+w_main(int argc, char *argv[])
 {
 	struct kinfo_proc *kp;
 	struct kinfo_proc *kprocbuf;
@@ -324,7 +326,7 @@
 			if (!usermatch)
 				continue;
 		}
-		if ((ep = calloc(1, sizeof(struct entry))) == NULL)
+		if ((ep = calloc(1, sizeof(struct utmp_entry))) == NULL)
 			errx(1, "calloc");
 		*nextp = ep;
 		nextp = &ep->next;
@@ -489,7 +491,7 @@
 	}
 	/* sort by idle time */
 	if (sortidle && ehead != NULL) {
-		struct entry *from, *save;
+		struct utmp_entry *from, *save;
 
 		from = ehead;
 		ehead = NULL;
